File: baekjoon/DataStructure/001-10868-최솟값.js
const [nm, ...input] = require('fs').readFileSync('z.in').toString().trim().split('\n');
const [N, M] = nm.split(' ').map(Number);
const arr = new Array(N);
for (let i = 0; i < N; i++) {
    arr[i] = Number(input[i]);
}

const MAX = Number.MAX_SAFE_INTEGER;
const width = 2 ** Math.ceil(Math.log2(arr.length));
const tree = new Array(2 * width);

function initTree(start = 0, end = arr.length-1, node = 1) {
    if (start == end) {
        tree[node] = arr[start];
        return tree[node];
    }

    let mid = parseInt((start + end) / 2);
    let leftVal = initTree(start, mid, node * 2);
    let rightVal = initTree(mid+1, end, node * 2 + 1);
    tree[node] = leftVal < rightVal ? leftVal : rightVal;
    return tree[node];
}

function getMin(left, right, start = 0, end = arr.length-1, node = 1) {
    if (start > right || end < left) return MAX;
    if (left <= start && end <= right) return tree[node];

    let mid = parseInt((start + end) / 2);
    let leftVal = getMin(left, right, start, mid, node * 2);
    let rightVal = getMin(left, right, mid+1, end, node * 2 + 1);
    return leftVal < rightVal ? leftVal : rightVal;
}

initTree();

let a, b;
for (let i = N; i < N + M; i++) {
    [a, b] = input[i].split(' ').map(Number);
    console.log(getMin(a-1, b-1));
}

File: baekjoon/DataStructure/001-10868-최솟값.py
import sys, math

MAX =  10 ** 9 + 1
MIN = 0

def init_tree(arr, tree, start, end, node = 1):
    if start == end:
        tree[node] = arr[start]
        return tree[node]

    mid = (start + end) // 2
    left_val = init_tree(arr, tree, start, mid, node * 2)
    right_val = init_tree(arr, tree, mid+1, end, node * 2 + 1)
    min_val = left_val if left_val < right_val else right_val
    tree[node] = min_val
    return tree[node]


def get_val(tree, left, right, start, end, node):
    if right < start or left > end:
        return MAX
    if left <= start and right >= end:
        return tree[node]

    mid = (start + end) // 2
    left_val = get_val(tree, left, right, start, mid, node * 2)
    right_val = get_val(tree, left, right, mid+1, end, node * 2 + 1)
    min_val = left_val if left_val < right_val else right_val
    return min_val


def main():
    input = sys.stdin.readline
    N, M = map(int, input().split())
    arr = [0] + [int(input()) for _ in range(N)]
    width = 2 ** math.ceil(math.log2(len(arr)))
    tree = [MAX for _ in range(2 * width)]

    init_tree(arr, tree, 1, N, 1)
    
    for _ in range(M):
        start, end = map(int, input().split())
        print(get_val(tree, start, end, 1, N, 1))


if __name__ == '__main__':
    main()